enumerable_methods.rb

```diff
@@ -0,0 +1,127 @@
+# rubocop: disable Metrics/ModuleLength
+# rubocop: disable Metrics/CyclomaticComplexity
+# rubocop: disable Metrics/PerceivedComplexity
+module Enumerable
+  def my_each
+    arr = self if self.class == Array
+    arr = flatten if self.class == Hash
+    arr = to_a if self.class == Range
+
+    count = 0
+    while count < arr.length
+      yield(arr[count], count)
+      count += 1
+    end
+    self
+  end
+
+  def my_each_with_index
+    my_each { |elem, count| yield(elem, count) }
+  end
+
+  def my_select
+    selected = []
+    my_each { |elem| selected << elem if yield(elem) }
+    selected
+  end
+
+  def my_all(arg = nil)
+    if block_given?
+      my_each { |elem| return false if yield(elem) != true }
+      return true
+    end
+
+    if arg.nil?
+      my_each { |elem| return false if elem.nil? || elem == false }
+    elsif arg.is_a? Class
+      my_each { |elem| return false if elem.class != arg }
+    elsif arg.class == Regexp
+      my_each { |elem| return false unless arg.match(elem) }
+    else
+      my_each { |elem| return false if elem != arg }
+    end
+    true
+  end
+
+  def my_any(arg = nil)
+    if block_given?
+      my_each { |elem| return true if yield(elem) == true }
+      return false
+    end
+
+    if arg.nil?
+      my_each { |elem| return true if elem.nil? || elem == true }
+    elsif arg.is_a? Class
+      my_each { |elem| return true if elem.class == arg }
+    elsif arg.class == Regexp
+      my_each { |elem| return true if arg.match(elem) }
+    else
+      my_each { |elem| return true if elem == arg }
+    end
+    false
+  end
+
+  def my_none(arg = nil)
+    arr = self
+    if block_given?
+      !arr.my_any { |elem| yield(elem) }
+    else
+      !arr.my_any(arg)
+    end
+  end
+
+  def my_count(num = nil)
+    arr = self
+    return arr.length if !block_given? && !num
+
+    return arr.my_select { |elem| elem == num }.length if num
+
+    arr.my_select { |elem| yield(elem) }.length
+  end
+
+  def my_map
+    arr = []
+    self_arr = self
+    self_arr.my_each do |elem|
+      arr << yield(elem)
+    end
+    arr
+  end
```

Comment:
```rb
2.6.5 :050 > array
 => [0, 4, 6, 0, 5, 6, 7, 0, 3, 7, 1, 8, 3, 2, 5, 1, 3, 3, 5, 5, 4, 7, 1, 6, 7, 6, 4, 6, 6, 6, 5, 1, 0, 7, 0, 1, 6, 2, 7, 8, 0, 7, 5, 4, 6, 1, 0, 7, 3, 4, 4, 6, 6, 1, 4, 2, 1, 4, 5, 1, 7, 8, 8, 7, 2, 2, 1, 2, 0, 7, 2, 4, 2, 3, 1, 0, 6, 5, 8, 2, 0, 4, 8, 3, 1, 2, 7, 7, 7, 3, 5, 3, 2, 1, 5, 6, 2, 0, 2, 6] 
2.6.5 :051 > array.map.is_a?(Enumerator)
 => true 
2.6.5 :052 > array.my_map.is_a?(Enumerator)
Traceback (most recent call last):
        9: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `<main>'
        8: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `eval'
        7: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `<main>'
        6: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `load'
        5: from /home/maelfosso/.rvm/gems/ruby-2.6.5/gems/irb-1.2.4/exe/irb:11:in `<top (required)>'
        4: from (irb):52
        3: from enumerable_methods.rb:85:in `my_map'
        2: from enumerable_methods.rb:12:in `my_each'
        1: from enumerable_methods.rb:86:in `block in my_map'
LocalJumpError (no block given (yield))
2.6.5 :053 >
```

Reply:
```rb
2.6.5 :055 > array
 => [0, 4, 6, 0, 5, 6, 7, 0, 3, 7, 1, 8, 3, 2, 5, 1, 3, 3, 5, 5, 4, 7, 1, 6, 7, 6, 4, 6, 6, 6, 5, 1, 0, 7, 0, 1, 6, 2, 7, 8, 0, 7, 5, 4, 6, 1, 0, 7, 3, 4, 4, 6, 6, 1, 4, 2, 1, 4, 5, 1, 7, 8, 8, 7, 2, 2, 1, 2, 0, 7, 2, 4, 2, 3, 1, 0, 6, 5, 8, 2, 0, 4, 8, 3, 1, 2, 7, 7, 7, 3, 5, 3, 2, 1, 5, 6, 2, 0, 2, 6] 
2.6.5 :056 > my_proc = proc {|num| num > 10 }
2.6.5 :057 > array.my_map(my_proc) {|num| num < 10 } == array.map(&my_proc)
Traceback (most recent call last):
        7: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `<main>'
        6: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `eval'
        5: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `<main>'
        4: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `load'
        3: from /home/maelfosso/.rvm/gems/ruby-2.6.5/gems/irb-1.2.4/exe/irb:11:in `<top (required)>'
        2: from (irb):57
        1: from enumerable_methods.rb:82:in `my_map'
ArgumentError (wrong number of arguments (given 1, expected 0))
2.6.5 :058 > 
```

Reply:
`my_map` is returning `LocalJumpError` instead of `true` when passing Enumerator, as in the screenshot below
![my_map](https://user-images.githubusercontent.com/46388113/89061000-7eb8ad80-d353-11ea-9d41-1b5a119686d6.png)